torch/csrc/jit/passes/create_autodiff_subgraphs.cpp

```diff
@@ -1,257 +1,293 @@
-  // NB: the block's param and return nodes are not in the dependency graph.
-  for (Node * node : block->nodes()) {
-    node_to_vertex[node] = dag.newVertex(node);
+  // Combine the nodes in two groups together. The nodes will end up in
+  // `consumerGroup`, and `producerGroup` will be deleted.
+  void mergeGroups(Node* consumerGroup, Node* producerGroup) {
+    // Extract the nodes in `producerGroup` into the outer graph
+    const auto nodes = unmergeGroup(producerGroup);
+    // Then merge them into `consumerGroup`
+    for (auto it = nodes.rbegin(); it != nodes.rend(); ++it) {
+      mergeNodeIntoGroup(consumerGroup, *it);
+    }
   }
-  for (auto * node : block->nodes()) {
-    for (auto * v : node->outputs()) {
-      for (auto & use : v->uses()) {
-        // [Determine data dependencies]
-        // Consider the following code:
-        //     y = f(x)
-        //     if k:
-        //        w += y
-        //     z = g(y)
-        // This produces a dependency graph with 3 vertices:
-        // (0: f)   (1: if k ...)   (2: g)
-        // We need to peek into the if Node* to determine its data dependencies
-        // (the body depends on the output of f, so Vertex 1 depends on Vertex 0).
-        // For each Use of y, we find an owning node of y that is a part of the
-        // dependency graph (in this case, the Vertex containing the if Node*)
-        // and then record the dependency.
-        auto * owning_node = use.user;
-        if (owning_node == block->return_node()) {
-          // The return node is not in the dag. Carry on.
-          continue;
-        }
-        while (true) {
-          auto search = node_to_vertex.find(owning_node);
-          if (search == node_to_vertex.end()) {
-            owning_node = owning_node->owningBlock()->owningNode();
-            JIT_ASSERT(owning_node != nullptr);
-            continue;
-          }
-          // NB: DynamicDAG is a simple graph (no multi-edges).
-          // addEdge is a no-op if the edge already exists.
-          dag.addEdge(node_to_vertex[node], search->second);
-          break;
+
+  // Merge node `n` into `group`'s subgraph
+  Node* mergeNodeIntoGroup(Node* group, Node* n) {
+    JIT_ASSERT(group->kind() == prim::DifferentiableGraph);
+    JIT_ASSERT(n->kind() != prim::DifferentiableGraph);
+
+    auto& subgraph = getSubgraph(group);
+
+    // Map from values in the surrounding graph to inputs in the subgraph
+    std::unordered_map<Value*, Value*> inputsMap;
+
+    JIT_ASSERT(group->inputs().size() == subgraph.inputs().size());
+    size_t i = 0;
+    for (auto input : group->inputs()) {
+      inputsMap[input] = subgraph.inputs()[i];
+      i++;
+    }
+
+    // Add n's inputs to the group's input list if we don't already have them
+    WithInsertPoint guard(*subgraph.nodes().begin());
+    for (auto input : n->inputs()) {
+      if (inputsMap.count(input) == 0) {
+        // Clone constants inside the subgraph instead of referencing them, to
```

Comment:
Why are we cloning constants into the subgraph? I thought constants weren't differentiable

Reply:
I just copypasta'd the original mergeNodes implementation here. It came from this PR: https://github.com/pytorch/pytorch/pull/11809

Reply:
Got it, constants are only cloned if they're inputs of nodes that that are going into the differentiable group. Thanks for the link!